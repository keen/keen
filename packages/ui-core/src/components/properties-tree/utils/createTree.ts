const getFromPath = (
  dataObject: Record<string, any>,
  selector: (string | number)[]
) =>
  selector.reduce((acc, key) => {
    if (acc === null) return dataObject[key];
    if (typeof acc === 'object' && acc !== null) return acc[key];
    return acc;
  }, null);

const setValue = (obj: Record<string, any>, path: string, value: string) => {
  const keys = path.split('.');
  let objectCopy = obj;

  const isPathExist = getFromPath(objectCopy, keys);

  while (keys.length - 1) {
    const key = keys.shift();
    if (!(key in objectCopy)) objectCopy[key] = {};
    objectCopy = objectCopy[key];
  }

  const key = isPathExist ? `${keys[0]} (node)` : keys[0];
  objectCopy[key] = [path, value];
};

/**
 * Creates tree from schema properties.
 *
 * @NOTES: reverse() is applied to support schema inconsistenty
 * and properties of list type
 *
 * @param properties - schema properties
 * @return tree object
 *
 */
const createTree = (properties: Record<string, any>) =>
  Object.keys(properties)
    .sort()
    .reverse()
    .reduce((acc, key) => {
      const propertyType = properties[key];
      setValue(acc, key, propertyType);
      return acc;
    }, {});

export default createTree;
