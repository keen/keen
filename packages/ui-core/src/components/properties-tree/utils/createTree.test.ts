import createTree from './createTree';

test('creates JSON tree', () => {
  const records = {
    'category.details.gender': 'string',
    'category.id': 'string',
    'category.details.name': 'string',
    age: 'number',
  };

  expect(createTree(records)).toMatchSnapshot();
});

test('creates JSON tree for conflicted nodes', () => {
  const records = {
    category: 'null',
    'category.details.gender': 'string',
    'category.id': 'string',
    'category.details.name': 'string',
    age: 'number',
  };

  expect(createTree(records)).toMatchSnapshot();
});

test.only('creates JSON tree for list in schema', () => {
  const records = {
    category: 'null',
    'category.details': 'list',
    'category.details.gender': 'string',
    'category.details.id': 'num',
    'category.details.0.id': 'num',
    'category.details.0.name': 'num',
    'category.details.1.id': 'num',
    'category.details.2.name': 'num',
  };

  expect(createTree(records)).toMatchSnapshot();
});
