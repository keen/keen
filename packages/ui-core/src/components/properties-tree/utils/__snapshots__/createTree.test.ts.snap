// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates JSON tree 1`] = `
Object {
  "age": Array [
    "age",
    "number",
  ],
  "category": Object {
    "details": Object {
      "gender": Array [
        "category.details.gender",
        "string",
      ],
      "name": Array [
        "category.details.name",
        "string",
      ],
    },
    "id": Array [
      "category.id",
      "string",
    ],
  },
}
`;

exports[`creates JSON tree for conflicted nodes 1`] = `
Object {
  "age": Array [
    "age",
    "number",
  ],
  "category": Object {
    "details": Object {
      "gender": Array [
        "category.details.gender",
        "string",
      ],
      "name": Array [
        "category.details.name",
        "string",
      ],
    },
    "id": Array [
      "category.id",
      "string",
    ],
  },
  "category (node)": Array [
    "category",
    "null",
  ],
}
`;

exports[`creates JSON tree for list in schema 1`] = `
Object {
  "category": Object {
    "details": Object {
      "0": Object {
        "id": Array [
          "category.details.0.id",
          "num",
        ],
        "name": Array [
          "category.details.0.name",
          "num",
        ],
      },
      "1": Object {
        "id": Array [
          "category.details.1.id",
          "num",
        ],
      },
      "2": Object {
        "name": Array [
          "category.details.2.name",
          "num",
        ],
      },
      "gender": Array [
        "category.details.gender",
        "string",
      ],
      "id": Array [
        "category.details.id",
        "num",
      ],
    },
    "details (node)": Array [
      "category.details",
      "list",
    ],
  },
  "category (node)": Array [
    "category",
    "null",
  ],
}
`;
